vowpalwabbit/core/src/reductions/epsilon_decay.cc

```diff
@@ -26,6 +28,126 @@ namespace epsilon_decay
 {
 float decayed_epsilon(uint64_t update_count) { return static_cast<float>(std::pow(update_count + 1, -1.f / 3.f)); }
 
+void epsilon_decay_data::update_weights(VW::LEARNER::multi_learner& base, VW::multi_ex& examples)
+{
+  auto model_count = static_cast<int64_t>(_scored_configs.size());
+  CB::cb_class logged{};
+  uint64_t labelled_action = 0;
+  const auto it =
+      std::find_if(examples.begin(), examples.end(), [](VW::example* item) { return !item->l.cb.costs.empty(); });
+  if (it != examples.end())
+  {
+    logged = (*it)->l.cb.costs[0];
+    labelled_action = std::distance(examples.begin(), it);
+  }
+
+  const float r = -logged.cost;
+  auto& ep_fts = examples[0]->_reduction_features.template get<VW::cb_explore_adf::greedy::reduction_features>();
+  // Process each model, then update the upper/lower bounds for each model
+  for (int64_t i = 0; i < model_count; ++i)
+  {
+    if (!_constant_epsilon)
+    { ep_fts.epsilon = VW::reductions::epsilon_decay::decayed_epsilon(_scored_configs[i][i].update_count); }
+    if (!base.learn_returns_prediction) { base.predict(examples, _weight_indices[i]); }
+    base.learn(examples, _weight_indices[i]);
+    for (const auto& a_s : examples[0]->pred.a_s)
+    {
+      if (a_s.action == labelled_action)
+      {
+        const float w = (logged.probability > 0) ? a_s.score / logged.probability : 0;
+        for (int64_t j = 0; j <= i; ++j) { _scored_configs[i][j].update(w, r); }
+        break;
+      }
+    }
+  }
+}
+
+// Promote model and all those lower with distance swap_dist
+void epsilon_decay_data::promote_model(int64_t model_ind, int64_t swap_dist)
```

Comment:
nit: `model_ind` and `swap_dist` should probably be unsigned ints. I dont think negative values make sense or necessarily even work in this function?

Reply:
model_ind can become negative in one of the loops, I kept it all int so all the types match